src/Adapter/Carrier/Repository/CarrierRepository.php

```diff
@@ -353,4 +353,40 @@ public function getLastPosition(): int
             ->fetchOne()
         ;
     }
+
+    /**
+     * Returns a mapping of product IDs to their available carriers.
+     *
+     * @param int[] $productIds list of product IDs
+     *
+     * @return array<int, array<int, array{id_carrier: int, name: string}>>
+     *
+     * An associative array where the key is the product ID and the value is an array of carriers.
+     * Each carrier is represented as an associative array with keys:
+     *  - id_carrier: The carrier ID.
+     *  - name: The carrier name.
+     */
+    public function findCarriersByProductIds(array $productIds, ShopId $shopId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb->select('pc.id_product AS product_id, c.id_carrier AS id_carrier, c.name')
```

Comment:
I guess you kept the column for the `where` clause, but it seems at least the alias is useless.

```suggestion
        $qb->select('pc.id_product, c.id_carrier AS id_carrier, c.name')
```